### DISCUSSION TITLE
Balance between table and database user defined function

### DISCUSSION THREAD

**Discussion by sharkAndshark** [maplibre/martin#431] [Ideas]
Hi guys.Here some very naive thoughts.If it's not Ok,just ignore.

There are two implementations of source for PostgreSQL : table/view, database function.If user want something a little beyond table,like tile grid  4326,filtering by some condition,cross tables/schema query...Choices they have:

* Create a view if it can solve the problem 
* Asking help from database function 

##  Parametrized Custom sql source 
Could we introduce another implementation of PostgreSQL source ,like:  

`custom_sql:  ……. ST_TileEnvelope(@z, @x,@y) ….
  WHERE geom && ST_TileEnvelope(@z,@x,@y)…. `

And Martin will inject z,x,y (or even other custom parameters from front end query )to sql in runtime.

**Comment by nyurik** [comment]
I was thinking about it too - essentially the usecase is a read-only database (no matter how unlikely), and a SQL statement that is used to generate a tile.  We can use the same approach I did in the [tilelive-pgquery](https://github.com/nyurik/tilelive-pgquery/blob/main/lib/PgQuery.js) - where a user-supplied query is examined (by asking DB to parse it), and made into a data source.  I'm all for prototyping it if you want. This probably should have been an issue rather than a discussion, but either one is ok.
